### ISSUE TITLE
toTry() in Value invokes get() if Value is not instance of Try

### ISSUE THREAD

**Issue by Opalo** [javaslang/javaslang#1875]
Hi,

Today while working with `Option` and mapping it to `Try` with `toTry` method I noticed (after seeing `NoSuchElementException` stacktrace in console) that this method invokes `get()` on `Value` if the given `Value` is not instance of `Try`. Unfortunately how it's implemented now makes impossible to process given `Value` fluently. IMO `toTry` method should return `Try` (if it's an instance of `Try` itself) or invoke `get()` if `Value` is present or return `Try.failure` if there is no `Value` present.  Please share your thoughts. I'm happy to provide a PR if necessary.

Below, code taken from `Value`:

```
    default Try<T> toTry() {
        if (this instanceof Try) {
            return (Try<T>) this;
        } else {
            return Try.of(this::get);
        }
    }
```

**Comment by danieldietrich** [comment]
I think it currently works fine and a fluent processing is possible:

| Source | Result |
| --- | --- |
| Some("hi").toTry() | Success("hi") |
| None().toTry() | Failure(NoSuchElementException) |

There are also conversion methods that allow to provide a custom exception in the empty/failure case:

| Source | Result |
| --- | --- |
| Some("hi").toTry(() -> new Error("oh")) | Success("hi") |
| None().toTry(() -> new Error("oh")) | Failure(Error("oh")) |

@Opalo Do you have an example where a fluent processing is not possible? Which of the above cases are not correct in you opinion?

**Comment by Opalo** [comment]
Hi @danieldietrich! 

Thanks for your reply and yes I've an example where in my opinion javaslang does not behave in a way I'm expecting - (no offence, I really do consider javaslang as a great library :) )

Have a look at the following piece of code:

```
    public void setImage(List<FileUpload> image) {
        if (image != null) {
            Option
                .of(image)
                .filter(fileUploads -> fileUploads.size() == 1)
                .map(fileUploads -> fileUploads.get(0))
                .map(fileUpload -> Tuple.of(Try.of(fileUpload::getInputStream), getExtension(fileUpload.getClientFileName())))
                .map(tuple -> Tuple.of(tuple._1, getImageFilename(tuple._2)))
                .map(tuple -> Tuple.of(tuple._1, getImagePath(tuple._2)))
                .toTry()
                .andThenTry(tuple -> deleteIfExists(tuple._2))
                .andThenTry(tuple -> copy(tuple._1.get(), tuple._2))
                .onSuccess(tuple -> log.info("Image: {} for lawyer: {} uploaded successfully", tuple._2, editable.getEmail()))
                .onFailure(err -> {
                    log.error("Error while uploading image for lawyer: {}, err: {}", editable.getEmail(), err);
                    throw new IllegalStateException(String.format("Error while uploading file for lawyer: %s", editable.getEmail()));
                });
        }
    }
```

This is an extract from wicket model - it's used for file uploading. Here's also an important excerpt from the docs:

> The model of this component is reset with null at the end of the request because FileUpload instances do not survive across requests since the input streams they point to will be closed. Because of this, the FileUpload instance should be processed within the same request as the form containing it was submitted.

TL;DR it means that `setImage` method will be invoked twice. First time with an argument and with `null` for the second time. I thought that wrapping the argument in an `Option` and then processing it will be sufficient, however when the method is called with `null` for the second time `toTry()` crashes with `NoSuchElementException`.  I'd rather expect it to become an instance of `Try.Failure` under the hood. 

I'm not sure if I've used javaslang properly here. In case of any objections just let me know. I'm happy to elaborate on this.

**Comment by danieldietrich** [comment]
@Opalo I see.

The `toTry()` method works correctly. There is no way to change it without breaking the Monad Laws in the way that the original state of the Option gets lost when converting it to Try.

More specifically an empty instance (None) needs to be converted to an empty instance (Failure).

You said:

> Unfortunately how it's implemented now makes impossible to process given Value fluently.

But that is not correct. If you comment out the `onFailure()` handler, the program does not throw. I.e. a fluent processing is possible.

The problem is that you throw the IllegalStateException. Or more specifically - you have 'good' errors (-> image is null) and 'bad' errors (-> like IOException on copy). But you don't distinguish between them.

Below you find an example that distinguishes between the errors. But I suggest to use a the `if (image != null)` you added.

```java
import javaslang.control.Option;
import javaslang.control.Try;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;

public class ToTry {

    public static void main(String[] args) {
        new ToTry().setImage(null);
    }

    static final Editable editable = new Editable();

    static final Logger log = new Logger();

    public void setImage(List<FileUpload> image) {
            Option
                    .of(image)
                    .filter(fileUploads -> fileUploads.size() == 1)
                    .map(fileUploads -> fileUploads.get(0))
                    .map(fileUpload -> Tuple.of(Try.of(fileUpload::getInputStream), getExtension(fileUpload.getClientFileName())))
                    .map(tuple -> Tuple.of(tuple._1, getImageFilename(tuple._2)))
                    .map(tuple -> Tuple.of(tuple._1, getImagePath(tuple._2)))
                    .toTry(ImageIsNullError::new)
                    .andThenTry(tuple -> deleteIfExists(tuple._2))
                    .andThenTry(tuple -> copy(tuple._1.get(), tuple._2))
                    .onSuccess(tuple -> log.info("Image: %s for lawyer: %s uploaded successfully", tuple._2, editable.getEmail()))
                    .onFailure(err -> {
                        if (!(err instanceof ImageIsNullError)) {
                            log.error("Error while uploading image for lawyer: {}, err: {}", editable.getEmail(), err);
                            throw new IllegalStateException(String.format("Error while uploading file for lawyer: %s", editable.getEmail()), err);
                        }
                    });
    }

    static String getImageFilename(String clientFilename) {
        return "";
    }

    static String getImagePath(String imageFilename) {
        return "";
    }

    static String getExtension(String fileName) {
        return "";
    }

    static void deleteIfExists(String fileName) {
    }

    static void copy(InputStream src, String dst) {
    }

    static class ImageIsNullError extends NullPointerException {
    }

    static class FileUpload {
        String getClientFileName() {
            return "";
        }

        InputStream getInputStream() throws IOException {
            return null;
        }
    }

    static class Editable {

        String getEmail() {
            return "a@bc.de";
        }
    }

    static class Logger {

        void info(String message, Object... args) {
            System.out.printf(message + "\n", args);
        }

        void error(String message, Object... args) {
            System.out.printf(message + "\n", args);
        }
    }
}
```

I think we can close this ticket.

**Comment by Opalo** [comment]
Thanks @danieldietrich for your time and valuable input! 👍

**Comment by danieldietrich** [comment]
@Opalo You are welcome. Please ask again on any questions!
